syntax = "proto3";

package data_format;

message MaxMinDoublePair {
  double max = 1;
  double min = 2;
}

message Location {
  double longitude_left = 1;
  double longitude_right = 2;
  double latitude_top = 3;
  double latitude_bottom = 4;
}

message Config {
  string name = 1;
  Location location = 2;
}

message Climate {
  enum ZoneType {
    TropicalWetAndDry = 0;
    TropicalWet = 1;
    DesertOrArid = 2;
    SteppeOrSemiArid = 3;
    SubtropicalHumid = 4;
    SubtropicalDrySummer = 5;
    SubtropicalDryWinter = 6;
    TemperateOceanic = 7;
    TemperateContinental = 8;
    TemperateWithHumidWinters = 9;
    TemperateWithDryWinters = 10;
    Boreal = 11;
    Polar = 12;
  }

  ZoneType climate_zone = 1;
  MaxMinDoublePair yearly_temperature = 2;
  MaxMinDoublePair yearly_rainfall = 3;
}

message Plant {
  enum Maturity {
    SEED = 0;
    SEEDLING = 1;
    JUVENILE = 2;
    MATURE = 3;
    OLD = 4;
  }

  string type_name = 1;
  uint32 health = 2;
  bool flowering = 3;
  int64 accumulated_gdd = 4;
  Maturity maturity = 5;
  double base_temperature = 6;
  repeated int64 gdd_thresholds = 7;
}

message Soil {
  enum Texture {
    CLAY = 0;
    SILT = 1;
    SAND = 2;
  }

  message Nutrients {
    double nitrogen = 1;
    double phosphorus = 2;
    double potassium = 3;
    double calcium = 4;
    double magnesium = 5;
    double sulfur = 6;
  }

  Texture texture = 1;
  double pH = 2;
  double salinity = 3;
  double organic_matter = 4;
  double water_content = 5;
}

message Terrain {
  message Cell {
    uint64 size = 1;
    oneof oneof_plant {
      Plant plant = 2;
    }
    Soil soil = 3;
  }

  message Row {
    repeated Cell cells = 1;
  }

  repeated Row tiles = 1;
}

message Weather {
  MaxMinDoublePair temperature = 1;
  double rainfall = 2;
}

message Environment {
  Config config = 1;
  Climate climate = 2;
  sint64 timestamp_epoch_count = 3;
  Terrain terrain = 4;
  Weather weather = 5;
}